google-cloud-firestore/lib/google/cloud/firestore/watch/listener.rb

```diff
@@ -0,0 +1,297 @@
+
+            new query, nil, query.client, init_listen_req, &callback
+          end
+
+          def initialize query, doc_ref, client, init_listen_req, &callback
+            @query = query
+            @doc_ref = doc_ref
+            @client = client
+            @init_listen_req = init_listen_req
+            @callback = callback
+
+            super() # to init MonitorMixin
+          end
+
+          def start
+            synchronize { start_listening! }
+            self
+          end
+
+          def stop
+            synchronize do
+              @stopped = true
+              @request_queue.push self if @request_queue
+            end
+          end
+
+          ##
+          # Whether the client has stopped listening for changes.
+          #
+          # @example
+          #   require "google/cloud/firestore"
+          #
+          #   firestore = Google::Cloud::Firestore.new
+          #
+          #   # Create a query
+          #   query = firestore.col(:cities).order(:population, :desc)
+          #
+          #   listener = query.listen do |snapshot|
+          #     puts "The query snapshot has #{snapshot.docs.count} documents "
+          #     puts "and has #{snapshot.changes.count} changes."
+          #   end
+          #
+          #   # Checks if the listener is stopped.
+          #   listener.stopped? #=> false
+          #
+          #   # When ready, stop the listen operation and close the stream.
+          #   listener.stop
+          #
+          #   # Checks if the listener is stopped.
+          #   listener.stopped? #=> true
+          #
+          def stopped?
+            synchronize { @stopped }
+          end
+
+          private
+
+          def send_callback query_snp
+            @callback.call query_snp
+          end
+
+          def start_listening!
+            # create new background thread to handle the stream's enumerator
+            @background_thread = Thread.new { background_run }
+          end
+
+          # @private
+          class RestartStream < StandardError; end
+
+          # rubocop:disable all
+
+          def background_run
+            # Don't allow a stream to restart if already stopped
+            return if synchronize { @stopped }
+
+            @backoff ||= { current: 0, delay: 1.0, max: 5, mod: 1.3 }
+
+            # Reuse inventory if one already exists
+            # Even though this uses an @var, no need to synchronize
+            @inventory ||= Inventory.new(@client, @query)
+            @inventory.restart
+
+            # Send stop if already running
+            synchronize do
+              @request_queue.push self if @request_queue
+            end
+
+            # Customize the provided initial listen request
+            init_listen_req = @init_listen_req.dup.tap do |req|
+              req.add_target.resume_token = String(@inventory.resume_token)
+              req.add_target.read_time = \
+                Convert.time_to_timestamp(@inventory.read_time)
```

Comment:
You should only one of the two - either `resume_token` or `read_time`. We only use the `resume_token` in the other clients.

Reply:
Okay. I’ll remove `read_time` in this cal.

Reply:
This has been removed.